modules/videoio/src/cap_ueye.cpp

```diff
@@ -0,0 +1,396 @@
+will be used. For any other number, the camera associated with that ID will be
+opened (c.f. IDS documentation for is_InitCamera).
+
+Images are double buffered in a ring buffer of size 2 (called 'image memory
+sequence' in the uEye SDK c.f. is_AddToSequence). The memory is locked on a
+'grab' call and copied and unlocked during 'retrieve'. The image queue provided
+in the uEye SDK is not used since it automatically locks the buffers when a new
+image arrives, which means the buffer can fill up when frames are retrieved too
+slowly.
+*/
+
+#include "precomp.hpp"
+
+#include <ueye.h>
+
+#include <array>
+#include <chrono>
+#include <cstdlib>
+#include <memory>
+#include <thread>
+
+namespace cv
+{
+namespace
+{
+struct image_buffer
+{
+    char* data;
+    INT id;
+};
+
+#define ASSERT_UEYE(expr) { UINT expr_result = expr; if(IS_SUCCESS != expr_result) CV_Error_(Error::StsAssert, ("%s %s %d: failed with code %u", #expr, __FILE__, __LINE__, expr_result)); }
+#define PRINT_ON_UEYE_ERROR( expr ) { UINT expr_result = expr; if(IS_SUCCESS != expr_result) std::cerr << #expr << " " << __FILE__ << " " << __LINE__ << ": failed with code " << expr_result << "\n"; }
+}
+
+struct VideoCapture_uEye CV_FINAL: public IVideoCapture
+{
+    int getCaptureDomain() CV_OVERRIDE
+    {
+        return cv::CAP_UEYE;
+    }
+
+    VideoCapture_uEye(int camera);
+
+    bool isOpened() const CV_OVERRIDE
+    {
+        return 255 != cam_id;
+    }
+
+    ~VideoCapture_uEye() CV_OVERRIDE
+    {
+        close();
+    }
+
+    double getProperty(int) const CV_OVERRIDE { return 0; }
+    bool setProperty(int, double) CV_OVERRIDE { return false; }
+    bool grabFrame() CV_OVERRIDE;
+    bool retrieveFrame(int, OutputArray) CV_OVERRIDE;
+
+    void close();
+    void start_camera();
+    void stop_camera();
+
+    void unlock_image_buffer();
+
+    HIDS cam_id = 255;
+    double fps;
+    int width;
+    int height;
+    int pitch;
+    std::array<image_buffer, 2> ring_buffer = {{{nullptr, 0}, {nullptr, 0}}};
+    char* locked_image = nullptr;
+};
+
+Ptr<IVideoCapture> create_ueye_camera(int camera)
+{
+    return cv::makePtr<VideoCapture_uEye>(camera);
+}
+
+VideoCapture_uEye::VideoCapture_uEye(int camera)
+{
+    try
+    {
+        CV_Assert(camera >= 0);
+        CV_Assert(camera < 255); // max camera id is 254
+        cam_id = static_cast<HIDS>(camera);
+        CV_LOG_DEBUG(NULL, "VIDEOIO(UEYE:" << cam_id << "): opening...");
+        ASSERT_UEYE(is_InitCamera(&cam_id, nullptr));
+
+        IS_INIT_EVENT init_event = {IS_SET_EVENT_FRAME, FALSE, FALSE};
+        ASSERT_UEYE(is_Event(cam_id, IS_EVENT_CMD_INIT, &init_event, sizeof(init_event)));
+        UINT frame_event = IS_SET_EVENT_FRAME;
+        ASSERT_UEYE(is_Event(cam_id, IS_EVENT_CMD_ENABLE, &frame_event, sizeof(frame_event)));
+        start_camera();
+    }
+    catch(const cv::Exception& e)
+    {
+        std::cerr << e.what() << '\n';
+        close();
```

Comment:
Why do you not want to propagate exception to caller?

Reply:
I was not sure when this is wanted in an `IVideoCapture`. From looking at the source, I gather throwing is OK in the constructor, but not in the member functions like `grabFrame`, `retrieve` etc.

Is that so?